common_test.go

```diff
@@ -81,12 +81,81 @@ func assertCipherSuiteParamsEquals(t *testing.T, a, b CipherSuiteParams) {
 
 func assertDeepEquals(t *testing.T, a, b interface{}) {
 	t.Helper()
-	assert(t, reflect.DeepEqual(a, b), fmt.Sprintf("%+v != %+v", a, b))
+	assertTrue(t, reflect.DeepEqual(a, b), fmt.Sprintf("%+v != %+v", a, b))
 }
 
 func assertSameType(t *testing.T, a, b interface{}) {
 	t.Helper()
 	A := reflect.TypeOf(a)
 	B := reflect.TypeOf(b)
-	assert(t, A == B, fmt.Sprintf("%s != %s", A.Name(), B.Name()))
+	assertTrue(t, A == B, fmt.Sprintf("%s != %s", A.Name(), B.Name()))
+}
+
+// Utilities for parametrized tests
+// Represents the configuration for a given test instance.
+type testInstanceState interface {
+	set(string, string)
+	get(string) string
+}
+
+// This complicated mixin stuff is in case we want to extend this to
+// pass other parameters that aren't srings. Arguably YAGNI.
+type testInstanceStateMixin struct {
+	params map[string]string
+}
+
+func (m *testInstanceStateMixin) set(k string, v string) {
+	m.params[k] = v
+}
+
+func (m *testInstanceStateMixin) get(k string) string {
+	return m.params[k]
+}
+
+func newTestInstanceStateMixin() testInstanceStateMixin {
+	return testInstanceStateMixin{make(map[string]string, 0)}
+}
+
+type testInstanceStateBase struct {
+	testInstanceStateMixin
+}
+
+func newTestInstanceStateBase() testInstanceState {
+	return &testInstanceStateBase{newTestInstanceStateMixin()}
+}
+
+// Helper function.
+func runParametrizedInner(t *testing.T, name string, state testInstanceState, inparams []testParameter, f parametrizedTest) {
+	next := inparams[1:]
+
+	param := inparams[0]
+	for _, p := range param.vals {
+		state.set(param.name, p)
+		var n string
+		if len(name) > 0 {
+			n = name + "/"
+		}
+		n = n + param.name + "=" + p
+
+		if len(next) == 0 {
+			t.Run(n, func(t *testing.T) {
+				f(t, n, state)
+			})
+			continue
+		}
+		runParametrizedInner(t, n, state, next, f)
+	}
+}
+
+// Nominally public API.
+type testParameter struct {
```

Comment:
It seems like instead of having this struct and a bunch of `[]testParameter`, you could just do something like:

`type testParams map[string][]string`

Reply:
You could indeed, but when I have this many []s I tend to get lost, so I'd prefer to leave as-is unless you object.